listenerutil/parse.go

```diff
@@ -387,3 +414,157 @@ func ParseSingleIPTemplate(ipTmpl string) (string, error) {
 		return "", fmt.Errorf("multiple addresses found (%q), please configure one", out)
 	}
 }
+
+// Header value consts
+const defaultStrictTransportSecurityHeader = "max-age=31536000; includeSubDomains"
+const defaultXContentTypeOptionsHeader = "nosniff"
+const defaultCacheControlHeader = "no-store"
+const defaultApiContentSecurityPolicyHeader = "default-src 'none'"
+const defaultUiContentSecurityPolicyHeader = "default-src 'none'; script-src 'self'; frame-src 'self'; font-src 'self'; connect-src 'self'; img-src 'self' data:*; style-src 'self'; media-src 'self'; manifest-src 'self'; style-src-attr 'self'; frame-ancestors 'self'"
+
+// Header names consts
+const contentSecurityPolicy = "Content-Security-Policy"
+const strictTransportSecurity = "Strict-Transport-Security"
+const xContentTypeOptions = "X-Content-Type-Options"
+const cacheControl = "Cache-Control"
+
+// parseCustomResponseHeaders takes raw config values for the "custom_ui_response_headers"
+// and "custom_api_response_headers". It makes sure the config entry is passed in as a map
+// of status code to a map of header name and header values. It verifies the validity of the
+// status codes, and header values. It also adds the default headers values for "Cache-Control",
+// "Strict-Transport-Security", "X-Content-Type-Options", and "Content-Security-Policy".
+func parseCustomResponseHeaders(responseHeaders interface{}, uiHeaders bool) (map[int]http.Header, error) {
+	h := make(map[int]http.Header)
+	// if responseHeaders is nil, we still should set the default custom headers
+	if responseHeaders == nil {
+		h[0] = http.Header{
+			strictTransportSecurity: {defaultStrictTransportSecurityHeader},
+			xContentTypeOptions:     {defaultXContentTypeOptionsHeader},
+			cacheControl:            {defaultCacheControlHeader},
+		}
+		if uiHeaders {
+			h[0][contentSecurityPolicy] = []string{defaultUiContentSecurityPolicyHeader}
+		} else {
+			h[0][contentSecurityPolicy] = []string{defaultApiContentSecurityPolicyHeader}
+		}
+		return h, nil
+	}
+
+	customResponseHeader, ok := responseHeaders.([]map[string]interface{})
+	if !ok {
+		return nil, fmt.Errorf("response headers were not configured correctly. Please make sure they're in a list of maps")
+	}
+
+	for _, crh := range customResponseHeader {
+		for statusCode, responseHeader := range crh {
+			headerValList, ok := responseHeader.([]map[string]interface{})
+			if !ok {
+				return nil, fmt.Errorf("response headers were not configured correctly. Please make sure they're in a list of maps")
+			}
+
+			status, err := convertStatusCode(statusCode)
+			if err != nil {
+				return nil, err
+			}
+
+			if len(headerValList) != 1 {
+				return nil, fmt.Errorf("invalid number of response headers exist")
+			}
+			headerValMap := headerValList[0]
+			headerVal, err := parseHeaders(headerValMap)
+			if err != nil {
+				return nil, err
+			}
+
+			h[status] = headerVal
+		}
+	}
+
+	// setting default headers
```

Comment:
This could be a bit more generic, a for loop over pairs of header names/defaults for example, but not that important.

Reply:
this is done in steps so that more specific status codes can overwrite less specific ones. ex, you can set a header on the default, but the same header on 200, and the 200 header will be returned instead of the default

and regarding preventing defaults, this is also possible by setting empty headers in your config file: https://github.com/hashicorp/go-secure-stdlib/pull/49/files#diff-462be8440ab1a6db8af3debb42b464e02433a564223b166d96dfbfb5702fdfb7R57-R58